### PR TITLE
First pass at loading and filtering table data for PDI

### PR DESCRIPTION
**Pull request by chris-allan**
This is a first pass at loading and filtering table data for PDI. It re-uses the existing implementation from OMERO.web for bulk annotation table loading.

The order of preference for bulk annotation discovery is now:

 * Project
 * Dataset
 * Plate

It could probably also be extended to Screen but that is really beyond the scope of what this PR is trying to do.

In addition to removing the obvious assumption that tables can only be associated with the SPW hierarchy, the assumption that Well columns are always the first column in the table has also been removed.

### CODE REVIEW

**Comment by will-moore** [comment]
Will try to test locally tomorrow...
One thought related to this PR:
When we get data in https://github.com/ome/omero-parade/blob/master/omero_parade/views.py#L153
we're actually loading the Table twice, because we call ```get_dataproviders``` AND then ```get_data```, both of which look-up the table and open it. 
This might have a small performance effect? *If* we wanted to avoid this, we could simply call ```get_data()``` for all data providers in turn and the  ```get_data``` implementation for each provider would simply return ```None``` if it didn't support the ```data_name```, e.g. if the column is not found in  ```index_column_index = column_names.index(index_column_name)```

File: table_filters/data_providers.py [hunk 1]
```diff
@@ -15,80 +15,110 @@
...
-    links = query_service.findAllByQuery(query, params, conn.SERVICE_OPTS)
-    shared_resources = conn.getSharedResources()
+
+def get_table(conn, objtype, objid):
     # Just use the first Table we find
     # TODO: handle multiple tables!?
-    file_id = links[0].child.file.id.val
-    table = shared_resources.openTable(OriginalFileI(file_id),
-                                       conn.SERVICE_OPTS)
+    data = _annotations(None, objtype, objid, conn=conn)['data'][0]
```
Comment (will-moore) [inline table_filters/data_providers.py]: It's possible to get a ```KeyError``` here if there's no annotations on the Dataset or an ```Index Out Of Bounds Exception``` if there's annotations but none with the ```NSBULKANNOTATIONS``` namespace ('data' is an empty list).

**Comment by will-moore** [comment]
Working fine for me, having added table to [dataset](http://web-dev-merge.openmicroscopy.org/webclient/?show=dataset-17701) and [project](http://web-dev-merge.openmicroscopy.org/webclient/?show=project-1851) on eel.

File: table_filters/data_providers.py (hunk 1 above)
Comment (chris-allan) [reply table_filters/data_providers.py]: In my rush to get this opened I've done a pretty bad job of exceptional condition handling overall. I'll make a full pass as part of fixing up the issues with columns which have spaces in their name.

**Comment by chris-allan** [comment]
Robustness now improved for both PDI and SPW. Bulk annotation data stored at the Screen level now supported along with columns which have non-alphanumeric characters in their name.

/cc @emilroz

**Comment by will-moore** [comment]
This is handling missing tables on Projects, Datasets etc just fine now.

**Comment by chris-allan** [comment]
Further robustness changes to handle column names with URL reserved sets of characters such as `#`, `?` and `/`.

**Comment by chris-allan** [comment]
Should be all ready now for final review and merging if you are still happy with the implementation @will-moore.

**Comment by will-moore** [comment]
Testing locally against a Plate with Table data works fine.
But when I connect my local OMERO.web to connect to eel, where I previously created a Table on Dataset and Project, something's not working...

screen shot 2018-03-23 at 13 07 22

When I select e.g. ```Table_Time``` from this list I get a 404 with 
http://localhost:4080/omero_parade/data/VGFibGVfVGltZQ==?project=1851

**Comment by chris-allan** [comment]
af3548259733f3cb4d4f884c3b3cad743122bf42 should handle that. I forgot that Base64 encoding could have characters that are not-alphanumeric.

**Comment by will-moore** [comment]
Yep - that's working now 👍

**Comment by chris-allan** [comment]
Definitely something to look at.

I am currently working on a follow up PR to this that will propose sparklines and some other metadata to help when choosing filter values that come from table data. That will move much of that calculation server side so we can make use of `numpy` niceties. Should help alleviate having this logic, and having to repeat it, in JavaScript.

File: table_filters/omero_filters.py [hunk 2]
```diff
@@ -74,7 +68,11 @@ def get_script(request, script_name, conn):
         f = """(function filter(data, params) {
             if (isNaN(params.count) || params.count == '') return true;
             var table_data = %s;
-            var rowIndex = table_data.Well.indexOf(data.wellId);
+            if (data.wellId) {
+                var rowIndex = table_data.Well.indexOf(data.wellId);
+            } else {
+                var rowIndex = table_data.Image.indexOf(data.id);
```
Comment (will-moore) [inline table_filters/omero_filters.py]: I recently found that the "Image" table column name doesn't match with "image" column name in ```populate_metadata.py``` https://github.com/openmicroscopy/openmicroscopy/blob/develop/components/tools/OmeroPy/src/omero/util/populate_metadata.py#L264 which I had used as a reference for my [batch_roi_export_to_table](https://github.com/ome/training-scripts/pull/9/files#diff-80c2e3fd72b532331e94c17950a506b4) script.
So we should probably align this, or at least document the convention that parade is expecting.

**Pull request merged**